open_spiel/games/gin_rummy.cc

```diff
@@ -0,0 +1,699 @@
+    // The non-knocking player does not have to arrange melds in a particular
+    // way to get under the knock card. Therefore we use kMaxPossibleDeadwood
+    // to ensure that all melds are legal.
+    std::vector<int> meld_ids = LegalMelds(hands_[cur_player_],
+                                           kMaxPossibleDeadwood);
+    for (int meld_id : meld_ids) {
+      int action_id = meld_id + kMeldActionBase;
+      legal_actions.push_back(action_id);
+    }
+  }
+  std::sort(legal_actions.begin(), legal_actions.end());
+  return legal_actions;
+}
+
+std::vector<Action> GinRummyState::WallLegalActions() const {
+  std::vector<Action> legal_actions;
+  legal_actions.push_back(kPassAction);
+  // Player can either pass or knock (if legal).
+  int deadwood = MinDeadwood(hands_[cur_player_], upcard_);
+  if (deadwood <= knock_card_) {
+    legal_actions.push_back(kKnockAction);
+  }
+  std::sort(legal_actions.begin(), legal_actions.end());
+  return legal_actions;
+}
+
+std::vector<std::pair<Action, double>> GinRummyState::ChanceOutcomes() const {
+  SPIEL_CHECK_TRUE(IsChanceNode());
+  std::vector<std::pair<Action, double>> outcomes;
+  const double p = 1.0 / stock_size_;
+  for (int card = 0; card < kNumCards; ++card) {
+    // This card is still in the deck, prob is 1/stock_size_.
+    if (deck_[card] != kInvalidCard) outcomes.push_back({card, p});
+  }
+  return outcomes;
+}
+
+std::string GinRummyState::ActionToString(Player player, Action action) const {
+  if (player == kChancePlayerId) {
+    return absl::StrCat("Chance outcome: ", CardString(action));
+  } else {
+    std::string action_str;
+    if (action < kNumCards) {
+      action_str = CardString(action);
+    } else if (action == kDrawUpcardAction) {
+      action_str = "Draw upcard";
+    } else if (action == kDrawStockAction) {
+      action_str = "Draw stock";
+    } else if (action == kPassAction) {
+      action_str = "Pass";
+    } else if (action == kKnockAction) {
+      action_str = "Knock";
+    } else if (action < kMeldActionBase + kNumMeldActions) {
+      std::vector<int> meld = int_to_meld.at(action - kMeldActionBase);
+      std::vector<std::string> meld_str = CardIntsToCardStrings(meld);
+      action_str = absl::StrJoin(meld_str, "");
+    } else {
+      SpielFatalError(absl::StrCat(
+          "Error in GinRummyState::ActionToString()."));
+    }
+    return absl::StrCat("Player: ", player, " Action: ", action_str);
+  }
+}
+
+std::string GinRummyState::ToString() const {
+  std::string rv;
+  absl::StrAppend(&rv, "\nKnock card: ", knock_card_);
+  absl::StrAppend(&rv, "\nPrev upcard: ", CardString(prev_upcard_));
+  absl::StrAppend(&rv, "\nRepeated move: ", repeated_move_);
+  absl::StrAppend(&rv, "\nPlayer turn: ", cur_player_);
+  std::vector<std::string> phase_str = {"Deal", "FirstUpcard", "Draw",
+                                        "Discard", "Knock", "Layoff",
+                                        "Wall", "GameOver"};
```

Comment:
might be worth hoisting out, and making a c array or a std::array.

Reply:
Incorporated into new commit.